packages/flutter/lib/src/foundation/logging.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:developer' as developer;
+
+/// A callback that, when evaluated, returns a log message.  Log messages must
+/// be encodable as JSON using `json.encode()`.
+typedef DebugLogMessageCallback = Object Function();
+
+/// Logs a message conditionally if the given identifying event [channel] is
+/// enabled (if `debugShouldLogEvent(channel)` is true).
+///
+/// Messages are obtained by evaluating [messageCallback] and must be encodable
+/// as JSON strings using `json.encode()`. In the event that logging is not
+/// enabled for the given [channel], [messageCallback] will not be evaluated.
+/// The cost of logging calls can be further mitigated at call sites by invoking
+/// them in a function that is only evaluated in profile or debug modes. For
+/// example,
+///
+/// ```dart
+/// profile(() {
+///   debugLogEvent(logGestures, () => <String, int> {
+///    'x' : x,
+///    'y' : y,
+///    'z' : z,
+///   });
+/// });
+///```
+///
+/// ignores logging entirely in release mode and no performance penalty is paid.
```

Comment:
That doesn't seem to be true... we still at a minimum pay for three method calls and some internal Set manipulation logic.

Reply:
we should probably wrap the entirety of all these debug* methods in `assert`s.

Reply:
In profile mode you should definitely just use the `Timeline` object. We already pay way too much overhead for that as it is, we shouldn't add more.

In practice, all the use cases that we've considered for this logging are things you'd do in debug mode, not in profile mode, as far as I can tell.

Reply:
What I meant by wrapping the methods was wrapping the actual implementation, not the calls, by the way. (I mean, the calls should be wrapped too, but that's another issue.) The point is to make sure that these methods _can't_ be called outside debug mode (that's why they have names starting with `debug`).